### PR TITLE
Add support for UDC for Apollo4P family

### PR DESCRIPTION
**Pull request by zeonchew**
This PR adds support for UDC for Ambiq Apollo4p family. The boards tested are: 
- apollo4p_evb
- apollo4p_blue_kxr_evb

Followings are the USB samples tested: 
- cdc_acm
- console
- hid-keyboard
- hid-mouse
- mass

### CODE REVIEW

File: drivers/usb/udc/udc_ambiq.c [hunk 1]
```diff
@@ -0,0 +1,915 @@
...
+K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_ambiq_event), CONFIG_UDC_AMBIQ_MAX_QMESSAGES,
+	      sizeof(void *));
+
+/* USB device controller access from devicetree */
+#define DT_DRV_COMPAT ambiq_usb
+
+/* Max Packet Size for USB endpoints */
+#define EP0_MPS   64U
+#define EP_FS_MPS 64U
+#define EP_HS_MPS 512U
```
Comment (tmon-nordic) [inline drivers/usb/udc/udc_ambiq.c:49]: Later the code sets the iso capability on the endpoints. Are the endpoints limited in hardware to these specific values and not the maximum allowed values?

File: samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#include <zephyr/dt-bindings/input/input-event-codes.h>
```
Comment (jfischer-no) [inline samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay]: Please add these to your board devicetree. My threshold for accepting overlays in samples is very high to keep the samples as clean as possible.

File: drivers/usb/udc/udc_ambiq.c [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+	uint32_t num_endpoints;
+	int speed_idx;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	void (*make_thread)(const struct device *dev);
+	void (*irq_enable_func)(const struct device *dev);
+	void (*irq_disable_func)(const struct device *dev);
+};
+
+/* Static variables */
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:70]: Please cleanup your code and remove all the unnecessary comments like `/* Static variables */`.

File: drivers/usb/udc/udc_ambiq.c [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+
+	return 0;
+}
+
+/* Enable Test Mode */
+static int udc_ambiq_test_mode(const struct device *dev, const uint8_t mode, const bool dryrun)
+{
+	LOG_DBG("Enter test mode.");
+
+	// TODO: Study and add TestMode support or remove test mode from interface
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c]: "Use C89-style single line comments, /*  */. The C99-style single line comment, //, is not allowed."

File: drivers/usb/udc/udc_ambiq.c [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+
+	__ASSERT(0, "Test mode not supported");
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:404]: Why is there ASSERT? ` * @retval -ENOTSUP Test mode is not supported` !

File: drivers/usb/udc/Kconfig.ambiq [hunk 6]
```diff
@@ -0,0 +1,34 @@
+ # Copyright (c) 2024 Ambiq Micro Inc.
```
Comment (jfischer-no) [inline drivers/usb/udc/Kconfig.ambiq]: What controller is it? Are there other vendors using the same controller design?

File: drivers/usb/udc/udc_ambiq.c [hunk 7]
```diff
@@ -25,6 +25,8 @@ enum udc_ambiq_event_type {
 	UDC_AMBIQ_EVT_HAL_IN_CMP,
 	/* Xfer request received via udc_ambiq_ep_enqueue API */
 	UDC_AMBIQ_EVT_XFER,
+	/* ZLP Pending to be transmitted */
+	UDC_AMBIQ_EVT_ZLP_PENDING,
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:29]: This commit should be squashed with the one that introduces the driver.
That you need a separate type for ZLP is suspicious.

File: drivers/usb/udc/udc_ambiq.c [hunk 8]
```diff
@@ -0,0 +1,915 @@
...
+		udc_ep_set_busy(dev, ep, false);
+		LOG_ERR("am_hal_usb_ep_xfer read(rx) failed(0x%02x), %d", ep, (int)status);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/* Handler for USB event callbacks from USB HAL */
+static void udc_evt_callback(am_hal_usb_dev_event_e eDevState)
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:156]: no cAmelCase

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/udc/udc_ambiq.c (hunk 1 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:49]: Hi @tmon-nordic, there is a limitation that the end of transaction is not able to be identified when OUT endpoint received data that is multiple of endpoint MPS, when the data count is lower than the triggered size. Hence, the safest way to expose this now is to limit it to 64 bytes for FS and 512 bytes for HS.

File: samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay (hunk 2 above)
Comment (zeonchew) [reply samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay]: Hi @jfischer-no, sorry for not knowing the rule for overlay. I have moved the configuration to board device tree. will take note the same for future PR.

File: drivers/usb/udc/udc_ambiq.c (hunk 3 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, it seems my comment style is too verbose, removed all the comments that is there just to describe what a struct/function/global-variable does. Please let me know if my style is still too verbose. Thanks.

File: drivers/usb/udc/udc_ambiq.c (hunk 4 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, sorry for missing out on the porting of test mode earlier. 
The actual test mode code is now uploaded. This should no longer be an issue. 😉

File: drivers/usb/udc/udc_ambiq.c (hunk 5 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, sorry for missing out on the porting of test mode earlier. 
The actual test mode code is now uploaded. This should no longer be an issue. 😉

File: drivers/usb/udc/Kconfig.ambiq (hunk 6 above)
Comment (zeonchew) [reply drivers/usb/udc/Kconfig.ambiq]: Hi @jfischer-no, the handling in this `udc_ambiq.c` is specific to Ambiq as we need to access the USB controller functionalities via Ambiq's USB HAL. I believe that the logic here is not reusable for USB controller in other vendor's chip.

File: drivers/usb/udc/udc_ambiq.c (hunk 7 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, Thanks for pointing this out. I was juggling on whether I should create a new event to trigger ZLP or just trigger the ZLP directly in complete event handler for IN transaction. 
Since you mentioned this as well, I think it's better to do the direct invocation instead. 
I have reverted to the said solution where there is no need for a separate `ZLP_PENDING` event

File: drivers/usb/udc/udc_ambiq.c (hunk 8 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, I have searched through and removed cAmelCase variable name throughout the file. Thanks for catching this 😉

File: samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay (hunk 2 above)
Comment (RichardSWheatley) [reply samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay]: @zeonchew a developer's personal threshold is not a rule in Zephyr.

The only rule is if the device is on the board, then add it to the board and not a sample/test overlay.
If it is not on-chip or onboard an overlay should always be acceptable practice.

Comment (zeonchew) [reply samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay]: Thanks @RichardSWheatley, Thanks for the explanation. 
In this case it is indeed better to put them into board as the declaration indeed belongs to the EVB boards. 
Thank you.

**Review by aaronyegx — DISMISSED** [review]
The boards and dts changes look good to me.

**Comment by zeonchew** [comment]
Hi @jfischer-no, I believe your comments above are addressed, please help to have a review again. Thank you 😉

**Review by tmon-nordic — DISMISSED** [review]

**Review by RichardSWheatley — DISMISSED** [review]

**Review by AlessandroLuo — DISMISSED** [review]

File: samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay (hunk 2 above)
Comment (jfischer-no) [reply samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay]: > The only rule is if the device is on the board, then add it to the board and not a sample/test overlay. If it is not on-chip or onboard an overlay should always be acceptable practice.

If there are no buttons on the board, the overlays would not be acceptable in any case!

File: drivers/usb/udc/udc_ambiq.c (hunk 1 above)
Comment (jfischer-no) [reply drivers/usb/udc/udc_ambiq.c:50]: Where is this limitation, in your shim driver, HAL driver, or controller?

File: boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts [hunk 9]
```diff
@@ -45,13 +46,18 @@
...
 			gpios = <&gpio0_31 17 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
 			label = "BTN0";
+			zephyr,code = <INPUT_KEY_0>;
+			status = "okay";
 		};
 		button1: button_1 {
 			gpios = <&gpio0_31 19 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
 			label = "BTN1";
+			zephyr,code = <INPUT_KEY_1>;
+			status = "okay";
```
Comment (jfischer-no) [inline boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:60]: The button changes should be in a separate commit, like all other board changes. Actually the commit sequence should look something like this: add input support to the boards, pull new manifest, dts changes, driver implementation commit, board changes to enable usb device controller.

File: drivers/usb/udc/udc_ambiq.c [hunk 10]
```diff
@@ -0,0 +1,925 @@
...
+	uint32_t num_endpoints;
+	int speed_idx;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	void (*make_thread)(const struct device *dev);
+	void (*irq_enable_func)(const struct device *dev);
+	void (*irq_disable_func)(const struct device *dev);
+};
+
+const static struct device *udc_ambiq_dev;
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:69]: How does it work if you have multiple instances?

File: drivers/usb/udc/udc_ambiq.c [hunk 11]
```diff
@@ -0,0 +1,925 @@
...
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		udc_ep_set_busy(dev, ep, false);
+		LOG_ERR("am_hal_usb_ep_xfer read(rx) failed(0x%02x), %d", ep, (int)status);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void udc_evt_callback(am_hal_usb_dev_event_e dev_state)
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:149]: Is this function called in ISR context?

File: drivers/usb/udc/udc_ambiq.c [hunk 12]
```diff
@@ -0,0 +1,925 @@
...
+
+	return 0;
+}
+
+static int udc_ambiq_ep_disable(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	__ASSERT_NO_MSG(ep_cfg);
+	am_hal_usb_ep_stall(priv->usb_handle, ep_cfg->addr);
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:353]: Why should disabled endpoint respond with stall handshake?

File: drivers/usb/udc/udc_ambiq.c [hunk 13]
```diff
@@ -0,0 +1,925 @@
...
+	}
+	udc_ep_set_busy(dev, ep, true);
+
+	/* buf equals NULL is used as indication of ZLP request */
+	if (buf == NULL) {
+		status = am_hal_usb_ep_xfer(priv->usb_handle, ep, NULL, 0);
+	} else {
+		status = am_hal_usb_ep_xfer(priv->usb_handle, ep, buf->data, buf->len);
+	}
+	if (status != AM_HAL_STATUS_SUCCESS) {
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:112]:
```c
	}

	if (status != AM_HAL_STATUS_SUCCESS) {
```

File: drivers/usb/udc/udc_ambiq.c [hunk 14]
```diff
@@ -0,0 +1,925 @@
...
+		buf = udc_buf_peek(dev, ep_addr);
+		if (buf == NULL) {
+			LOG_ERR("No buffer for ep 0x%02x", ep_addr);
+			udc_submit_event(dev, UDC_EVT_ERROR, -ENOBUFS);
+			return;
+		}
+		net_buf_add(buf, xfer_len);
+		evt.type = UDC_AMBIQ_EVT_HAL_OUT_CMP;
+	}
+	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:235]:
```c
		}

		net_buf_add(buf, xfer_len);
		evt.type = UDC_AMBIQ_EVT_HAL_OUT_CMP;
	}

	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
```

File: drivers/usb/udc/udc_ambiq.c [hunk 15]
```diff
@@ -0,0 +1,925 @@
...
+		break;
+	default:
+		return -EINVAL;
+	}
+	status = am_hal_usb_ep_init(priv->usb_handle, ep_cfg->addr, endpoint_type, ep_cfg->mps);
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		LOG_ERR("am_hal_usb_ep_init failed(0x%02x), %d", ep_cfg->addr, (int)status);
+		return -EIO;
+	}
+	LOG_DBG("Enable ep 0x%02x", ep_cfg->addr);
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:343]:
```c
	}

	status = am_hal_usb_ep_init(priv->usb_handle, ep_cfg->addr, endpoint_type, ep_cfg->mps);
	if (status != AM_HAL_STATUS_SUCCESS) {
		LOG_ERR("am_hal_usb_ep_init failed(0x%02x), %d", ep_cfg->addr, (int)status);
		return -EIO;
	}

	LOG_DBG("Enable ep 0x%02x", ep_cfg->addr);
```

File: drivers/usb/udc/udc_ambiq.c [hunk 16]
```diff
@@ -0,0 +1,925 @@
...
+	int ret = 0;
+	struct gpio_dt_spec usb_vddusb33_gpio = GPIO_DT_SPEC_GET(USB_DEV_NODE, vddusb33_gpios);
+	struct gpio_dt_spec usb_vddusb0p9_gpio = GPIO_DT_SPEC_GET(USB_DEV_NODE, vddusb0p9_gpios);
+
+	/* Enable USB IO */
+	ret = gpio_pin_configure_dt(&usb_vddusb33_gpio, GPIO_OUTPUT);
+	if (ret) {
+		return ret;
+	}
+	ret = gpio_pin_configure_dt(&usb_vddusb0p9_gpio, GPIO_OUTPUT);
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:473]:
```c
	}

	ret = gpio_pin_configure_dt(&usb_vddusb0p9_gpio, GPIO_OUTPUT);
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/udc/udc_ambiq.c (hunk 1 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:50]: I think it can be categorized as controller limitation.

File: drivers/usb/udc/udc_ambiq.c (hunk 10 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:69]: There is only 1 USB device controller in Ambiq chips. 
I believe that there will not be multiple instances for UDC for those chips.

File: drivers/usb/udc/udc_ambiq.c (hunk 11 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:149]: Yes, they are called in ISR context. 
Is there anything needs to be taken care since that is the case?

File: boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts (hunk 9 above)
Comment (zeonchew) [reply boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:60]: Hi @jfischer-no, the commit is now split with the suggested manner and sequence.

File: drivers/usb/udc/udc_ambiq.c (hunk 12 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, this is indeed a mistake. Thanks for catching this. It is corrected as follows: 
https://github.com/zephyrproject-rtos/zephyr/blob/ce491e0f5379c01cb5fb26bbbfe437e839b75a10/drivers/usb/udc/udc_ambiq.c#L353-L362

**Review by tmon-nordic — DISMISSED** [review]

File: drivers/usb/udc/udc_ambiq.c (hunk 10 above)
Comment (jfischer-no) [reply drivers/usb/udc/udc_ambiq.c:69]: Looks like bad HAD driver design. But variables below can be moved to udc_ambiq_data.

File: drivers/usb/udc/udc_ambiq.c (hunk 11 above)
Comment (jfischer-no) [reply drivers/usb/udc/udc_ambiq.c:150]: No udc_evt_callback() is fine. What else is called in the ISR context?

File: drivers/usb/udc/udc_ambiq.c [hunk 17]
```diff
@@ -0,0 +1,931 @@
...
+	am_hal_usb_intr_status_get(priv->usb_handle, &int_status[0], &int_status[1],
+				   &int_status[2]);
+	am_hal_usb_interrupt_service(priv->usb_handle, int_status[0], int_status[1], int_status[2]);
+}
+
+static int usb_power_rails_set(bool on)
+{
+#define USB_DEV_NODE DT_NODELABEL(zephyr_udc0)
+#if DT_NODE_HAS_STATUS(USB_DEV_NODE, okay) && DT_NODE_HAS_PROP(USB_DEV_NODE, vddusb33_gpios) &&    \
+	DT_NODE_HAS_PROP(USB_DEV_NODE, vddusb0p9_gpios)
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:468]: This is not correct. (Also, you cannot use `DT_NODELABEL(zephyr_udc0)` here).
Move vddusb33_gpio and vddusb0p9_gpio to `struct udc_ambiq_config`, use `GPIO_DT_SPEC_GET_OR(DT_DRV_INST(n), reset_gpios, {0})`;

File: boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts (hunk 9 above)
Comment (jfischer-no) [reply boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:60]: In the second commit, there are changes to the boards to use drivers that will be introduced in the following commit, which is not what I asked for.

Comment (zeonchew) [reply boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:60]: Hi @jfischer-no, looks like I did not understand correctly the intent of the original comments. 
I have changed the commit sequence to the following: 

1. Add user button to the boards
2. DTS binding
3. Driver implementation 
4. Board changes to enable USB device controller

File: drivers/usb/udc/udc_ambiq.c (hunk 11 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:150]: Hi @jfischer-no , `udc_ep0_setup_callback()` and `udc_ep_xfer_complete_callback()` is also called in the ISR context.

File: drivers/usb/udc/udc_ambiq.c (hunk 17 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Thanks @jfischer-no for the pointers. Added the change to use `GPIO_DT_SPEC_GET_OR` instead.

File: drivers/usb/udc/udc_ambiq.c (hunk 10 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Yes. You are right, that was indeed a bad driver design!
I have removed the dependency on `udc_ambiq_dev` of the functions in this driver, and moved all the variable into instance data structure `udc_ambiq_data` 

Thank you so much for pointing this out @jfischer-no

File: drivers/usb/udc/udc_ambiq.c [hunk 18]
```diff
@@ -0,0 +1,960 @@
...
+static void udc_ambiq_evt_callback(const struct device *dev, am_hal_usb_dev_event_e dev_state)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	switch (dev_state) {
+	case AM_HAL_USB_DEV_EVT_BUS_RESET:
+		/* enable usb bus interrupts */
+		am_hal_usb_intr_usb_enable(priv->usb_handle,
+					   USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
+						   USB_CFG2_SuspendE_Msk | USB_CFG2_ResetE_Msk);
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:160]:
```c
		am_hal_usb_intr_usb_enable(priv->usb_handle,
					   USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
					   USB_CFG2_SuspendE_Msk | USB_CFG2_ResetE_Msk);
```

File: drivers/usb/udc/udc_ambiq.c [hunk 19]
```diff
@@ -0,0 +1,960 @@
...
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	uint32_t int_status[3];
+
+	am_hal_usb_intr_status_get(priv->usb_handle, &int_status[0], &int_status[1],
+				   &int_status[2]);
+	am_hal_usb_interrupt_service(priv->usb_handle, int_status[0], int_status[1], int_status[2]);
+}
+
+static int usb_power_rails_set(const struct device *dev, bool on)
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:464]: The return value is never checked, maybe `static void usb_power_rails_set(const struct device *dev, bool on)`?

File: drivers/usb/udc/udc_ambiq.c [hunk 20]
```diff
@@ -0,0 +1,960 @@
...
+
+	/* power rails set */
+	ret = gpio_pin_set_dt(&cfg->usb_vddusb33_gpio, on == true ? 1 : 0);
+	if (ret) {
+		return ret;
+	}
+	ret = gpio_pin_set_dt(&cfg->usb_vddusb0p9_gpio, on == true ? 1 : 0);
+	if (ret) {
+		return ret;
+	}
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_ambiq.c:494]: Just `on`, bool is either 0 or 1 and can be promoted to any integer type.
```c
	/* power rails set */
	ret = gpio_pin_set_dt(&cfg->usb_vddusb33_gpio, on);
```

File: drivers/usb/udc/udc_ambiq.c [hunk 21]
```diff
@@ -0,0 +1,960 @@
...
+	bool ctrl_setup_recv_at_status_in;
+};
+
+struct udc_ambiq_config {
+	uint32_t num_endpoints;
+	int speed_idx;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	struct gpio_dt_spec usb_vddusb33_gpio;
+	struct gpio_dt_spec usb_vddusb0p9_gpio;
```
[the remark this answers is not in the document]
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, I think I have missed that point earlier. removed `usb_` from the 2 variable names

File: drivers/usb/udc/udc_ambiq.c (hunk 18 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:160]: clang-format keeps add a tab there. I have manually reverted the change by clang-format and push. Let's see whether the CI will let us through 😄

File: drivers/usb/udc/udc_ambiq.c (hunk 20 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c]: Hi @jfischer-no, understood and made the change. Will use implicit bool to integer casting in the future.

File: drivers/usb/udc/udc_ambiq.c (hunk 19 above)
Comment (zeonchew) [reply drivers/usb/udc/udc_ambiq.c:464]: Hi @jfischer-no, since the error status return might be useful to indicate that the power rail control is not successful, I have made the change to propagate the error to upper layer. Please help to have a look at the change and see whether it is okay.

**Review by RichardSWheatley — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]
